### PR TITLE
PassGenerator Tweaks, Tests, And Visualisation

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
- returned pair of pass and score instead of just pass
- removed pass score cutoff, callers can enforce it themselves
- made the pass generator convergence test more robust so it should always pass in CI
- added the functionality to the `PassGenerator` to set the passer robot id, so that we don't converge to a pass that is the passer robot passing to itself
- changed the `CanvasMessenger` to allow indepedent layer clearing and publishing
- added visualisation of the pass receive position gradient and the various passes to the PassGenerator
- added some more unit tests to the pass generator and hopefully made it more robust
- changed `run_ai` parameter to default to true (this seems reasonable until we get controls in the visualizer up and running)

### Testing Done
Unit tests, grsim testing.

### Resolved Issues
NA

### Length Justification
NA

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/ai/hl/stp/play/example_play.cpp [hunk 1]
```diff
@@ -23,6 +25,15 @@ bool ExamplePlay::invariantHolds(const World &world) const
...
     TacticCoroutine::push_type &yield, const World &world)
 {
+    AI::Passing::PassGenerator pass_generator(world, world.ball().position());
+
+    do {
+        pass_generator.setWorld(world);
+        pass_generator.setPasserPoint(world.ball().position());
+        std::this_thread::yield();
+        yield({});
+    } while(true);
```
Comment (garethellis0) [inline src/thunderbots/software/ai/hl/stp/play/example_play.cpp]: We should definitely remove this before merging, just have it here so people and pull at test out the changes.

File: src/thunderbots/software/ai/passing/pass_generator.h [hunk 2]
```diff
@@ -72,6 +79,16 @@ namespace AI::Passing
          */
         void setPasserPoint(Point passer_point);
 
+        /**
+         * Set the id of the robot performing the pass.
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.h:83]: Just to be explicit, indicate we assume this robot is on the friendly team

File: src/thunderbots/software/ai/passing/pass_generator.h [hunk 3]
```diff
@@ -108,8 +124,12 @@ namespace AI::Passing
 
         // Weights used to normalize the parameters that we pass to GradientDescent
         // (see the GradientDescent documentation for details)
+        // These weights are *very* roughly the step that gradient descent will take
+        // in each respective dimension for a single iteration. They are tuned to
+        // ensure passes converge as fast as possible, but are also as stable as
+        // possible
         static constexpr double PASS_SPACE_WEIGHT                          = 0.01;
-        static constexpr double PASS_TIME_WEIGHT                           = 0.1;
+        static constexpr double PASS_TIME_WEIGHT                           = 0.01;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/ai/passing/pass_generator.h]: I thought the `PASS_SPEED_WEIGHT` was going to be changed, not the `PASS_TIME_WEIGHT` to improve convergence?

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp [hunk 4]
```diff
@@ -7,20 +7,66 @@
...
+        {
+            prev_prev_score = prev_score;
+            prev_score      = curr_score;
+
+            std::this_thread::sleep_for(1s);
+            seconds_so_far++;
+
+            auto curr_pass_and_score = pass_generator->getBestPassSoFar();
+            curr_score               = curr_pass_and_score.second;
+            std::cout << curr_score << std::endl;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/ai/passing/pass_generator.cpp]: Probably don't want to leave printouts in the test

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp [hunk 5]
```diff
@@ -7,20 +7,66 @@
...
+        double prev_prev_score = 0;
+        do
+        {
+            prev_prev_score = prev_score;
+            prev_score      = curr_score;
+
+            std::this_thread::sleep_for(1s);
+            seconds_so_far++;
+
+            auto curr_pass_and_score = pass_generator->getBestPassSoFar();
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/ai/passing/pass_generator.cpp:53]: Do we not need to update the world for the pass_generator, or is this test invariant of time? If it is invariant of time, it should be documented as such.

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp [hunk 6]
```diff
@@ -29,54 +75,159 @@ class PassGeneratorTest : public testing::Test
...
-        EXPECT_LE(
-            (converged_pass_opt->receiverPoint() - new_pass_opt->receiverPoint()).len(),
-            0.2);
-        EXPECT_LE(converged_pass_opt->speed() - new_pass_opt->speed(), 0.2);
-        EXPECT_LE(abs((converged_pass_opt->startTime() - new_pass_opt->startTime())
-                          .getSeconds()),
-                  0.2);
+        auto [pass, score] = pass_generator->getBestPassSoFar();
+
+        std::cout << converged_pass << std::endl;
```
Comment (MathewMacDougall) [inline src/thunderbots/software/test/ai/passing/pass_generator.cpp]: Probably don't want printouts in the test

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp [hunk 7]
```diff
@@ -18,42 +18,57 @@ void CanvasMessenger::initializePublisher(ros::NodeHandle node_handle)
...
-    // Do not do anything if the time passed hasn't been
-    // long enough
-    if (elapsed_ms < DESIRED_PERIOD_MS)
-        return;
-
-    // Send a payload per layer of messages
-    for (const auto& layer_pair : this->layers_map)
+    std::lock_guard<std::mutex> layers_map_lock(layers_map_mutex);
+
+    //    // Limit rate of the message publishing
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: This should be implemented or the commented out code should be removed

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp [hunk 8]
```diff
@@ -18,42 +18,57 @@ void CanvasMessenger::initializePublisher(ros::NodeHandle node_handle)
...
 
-    // Clear shapes in layers of current frame/tick
-    this->clearAllLayers();
+        // Clear the layer
+        layer_pair->second = {};
+    }
 
-    // Update last published time
-    time_last_published = now;
+    //    // Send a payload per layer of messages
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: This should be implemented or removed

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp [hunk 9]
```diff
@@ -74,8 +89,11 @@ void CanvasMessenger::publishPayload(uint8_t layer, std::vector<Sprite> sprites)
     thunderbots_msgs::CanvasLayer new_layer;
     new_layer.data = payload;
 
-    // and publish
-    this->publisher.publish(new_layer);
+    // and publish if we have a valid ROS publisher
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Explain the scenario where we don't have a valid publisher (ie. unit tests without ROS running). Useful for context.

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp [hunk 10]
```diff
@@ -130,6 +148,45 @@ void CanvasMessenger::drawRectangle(Layer layer, Rectangle rectangle, Angle orie
...
+                                   const Rectangle& area, double min_val, double max_val,
+                                   Color min_color, Color max_color, int points_per_meter)
+{
+    for (int i = 0; i < area.width() * points_per_meter; i++)
+    {
+        for (int j = 0; j < area.height() * points_per_meter; j++)
+        {
+            Point p = area.swCorner() +
+                      Vector(0.5 / points_per_meter, 0.5 / points_per_meter) +
+                      Vector(i / (double)points_per_meter, j / (double)points_per_meter);
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: preferably use `static_cast`

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp [hunk 11]
```diff
@@ -130,6 +148,45 @@ void CanvasMessenger::drawRectangle(Layer layer, Rectangle rectangle, Angle orie
     drawSprite(layer, rectangle_sprite);
 }
 
+void CanvasMessenger::drawGradient(Layer layer, std::function<double(Point)> f,
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Can `f` be names more descriptively?

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp [hunk 12]
```diff
@@ -130,6 +148,45 @@ void CanvasMessenger::drawRectangle(Layer layer, Rectangle rectangle, Angle orie
...
+                      Vector(0.5 / points_per_meter, 0.5 / points_per_meter) +
+                      Vector(i / (double)points_per_meter, j / (double)points_per_meter);
+
+            // Get the value and clamp it appropriately
+            double val_at_p = std::clamp(f(p), min_val, max_val);
+
+            // Create the "pixel" in the gradient
+            Rectangle block(p - Vector(0.5 / points_per_meter, 0.5 / points_per_meter),
+                            p + Vector(0.5 / points_per_meter, 0.5 / points_per_meter));
+            //            Rectangle block(area.swCorner(), area.swCorner() +
```
Comment (MathewMacDougall) [inline src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Remove commented out code

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp [hunk 13]
```diff
@@ -37,17 +37,15 @@ std::unique_ptr<Intent> CherryPickTactic::calculateNextIntent(
     intent_coroutine::push_type& yield)
 {
     MoveAction move_action                     = MoveAction();
-    std::optional<AI::Passing::Pass> best_pass = pass_generator.getBestPassSoFar();
+    auto best_pass_and_score                   = pass_generator.getBestPassSoFar();
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/ai/hl/stp/tactic/cherry_pick_tactic.cpp:40]: We already set it (in the constructor for `CherryPickTactic`) and I don't see a case where we would have a cherry picker with no target region. At that point you should be using a `ReceiverTactic`.

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 14]
```diff
@@ -113,6 +133,57 @@ void PassGenerator::continuouslyGeneratePasses()
...
+    auto canvas_messenger = Util::CanvasMessenger::getInstance();
+
+    // Get field characteristics and the current time
+    world_mutex.lock();
+    Rectangle field_area(world.field().enemyCornerNeg(),
+                         world.field().friendlyCornerPos());
+    double field_length = world.field().length();
+    double field_width  = world.field().width();
+    Timestamp pass_zero_time =
+        world.ball().lastUpdateTimestamp() + Duration::fromSeconds(0.0);
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: Whoops! We actually don't use this. Removed!

File: src/thunderbots/software/ai/passing/pass_generator.cpp [hunk 15]
```diff
@@ -113,6 +133,57 @@ void PassGenerator::continuouslyGeneratePasses()
...
+    world_mutex.unlock();
+
+    auto best_pass_and_score      = getBestPassSoFar();
+    auto [best_pass, best_score]  = best_pass_and_score;
+    const auto objective_function = [&](Point p) {
+        try
+        {
+            double pass_speed =
```
[the remark this answers is not in the document]
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.cpp]: removed.

File: src/thunderbots/software/ai/passing/pass_generator.h (hunk 2 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.h:83]: specified.

File: src/thunderbots/software/ai/passing/pass_generator.h (hunk 3 above)
Comment (garethellis0) [reply src/thunderbots/software/ai/passing/pass_generator.h]: Oh whoops, yup, totally correct. Reverted change to `PASS_TIME_WEIGHT` and changed reduced `PASS_SPEED_WEIGHT` by a factor of 10.

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp (hunk 4 above)
Comment (garethellis0) [reply src/thunderbots/software/test/ai/passing/pass_generator.cpp]: Leaving this one here so i can see what's going on when CI inevitably causes issues for these tests, but yup, please leave this unresolved so I remember to remove them before merging.

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp (hunk 5 above)
Comment (garethellis0) [reply src/thunderbots/software/test/ai/passing/pass_generator.cpp:53]: It is invariant, added doc comment:
```
    // Test that we can converge to a stable pass in a fairly simple scenario.
    // It is difficult to update all the timestamps in the world that the pass generator
    // could use, so we don't, and hence this test does not really test convergence
    // of pass start time.
```

File: src/thunderbots/software/test/ai/passing/pass_generator.cpp (hunk 6 above)
Comment (garethellis0) [reply src/thunderbots/software/test/ai/passing/pass_generator.cpp]: See above comment, but yep, please leave unresolved for now.

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp (hunk 7 above)
Comment (garethellis0) [reply src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Implemented!

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp (hunk 8 above)
Comment (garethellis0) [reply src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Implemented

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp (hunk 9 above)
Comment (garethellis0) [reply src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Updated comment to:
```
    // and publish if we have a valid ROS publisher.
    // This check is important for cases where we're running this without running ROS,
    // such as in unit tests.
```

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp (hunk 10 above)
Comment (garethellis0) [reply src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Changed here and in a few other spots in this file.

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp (hunk 11 above)
Comment (garethellis0) [reply src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: Changed to `valueAtPoint`.

File: src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp (hunk 12 above)
Comment (garethellis0) [reply src/thunderbots/software/util/canvas_messenger/canvas_messenger.cpp]: removed

**Review by MathewMacDougall — APPROVED** [review]
LGTM :+1:

**Pull request merged**